File: algorithms/binary_search.java
// Binary Search is an efficient algorithm for finding a target value within a sorted array.
// It works by repeatedly dividing in half the portion of the list that could contain the target value, and comparing the target value to the middle element of the subarray.
// The algorithm eliminates the half in which the target value cannot lie in each iteration.

// Imagine you have a phonebook with names sorted alphabetically.
// You want to find a specific person's phone number.
// Instead of starting from the beginning and looking at each name one by one, you decide to open the phonebook roughly in the middle.
// If the name you're looking for comes after the middle name alphabetically, you repeat the process with the second half of the phonebook.
// Otherwise, you repeat the process with the first half.
// You keep repeating this process until you find the name you're looking for or determine it's not in the phonebook.

File: algorithms/binary_search.kt
// Binary Search is a searching algorithm that finds the position of a target value within a sorted array or list.
// It compares the target value to the middle element of the array and eliminates half of the remaining search interval based on that comparison.
// The process continues until the value is found or the interval is empty.

// Imagine you're looking up a word in a dictionary. Instead of going through each word from the beginning to the end,
// you'd typically open the dictionary to a middle page and see if your word would lie before or after that page.
// You then split the relevant half into another half, and continue until you find your word.
// This method is faster than a linear search through every page, and is essentially a real-life binary search.

fun binarySearch(arr: IntArray, target: Int): Int {
    var left = 0
    var right = arr.size - 1

    while (left <= right) {
        val mid = left + (right - left) / 2

        when {
            arr[mid] == target -> return mid
            arr[mid] < target -> left = mid + 1
            else -> right = mid - 1
        }
    }
    return -1
}

fun main() {
    val numbers = intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
    val target = 7

    val result = binarySearch(numbers, target)

    if (result != -1) {
        println("Element $target found at index $result")
    } else {
        println("Element $target not found in the array.")
    }
}
